src/urn.cc

```diff
@@ -275,8 +272,8 @@ urnHandleReply(void *data, StoreIOBuffer result)
 
     debugs(52, 3, "urnHandleReply: Called with size=" << result.length << ".");
 
-    if (EBIT_TEST(urlres_e->flags, ENTRY_ABORTED) || result.length == 0 || result.flags.error) {
```

Comment:
IIRC, the problem here was that when result.length == 0, we return early from urnHandleReply() without replying, which is performed in the same method below.

> If there is no more data to copy, we should not call storeClientCopy() below.

I agree. I think if result.length==0, store_status cannot be STORE_PENDING. For example, a similar mgr callback (see Mgr::StoreToCommWriter::NoteStoreCopied() and Mgr::StoreToCommWriter::noteStoreCopied()) has a Must about that:

```
    if (ioBuf.length > 0)
        scheduleCommWrite(ioBuf); // write received action results to client
    else
        Must(doneAll()); // otherwise, why would Store call us with no data?
```
If this is correct, all storeClientCopy() callbacks should insist that requirement. Note that a similar asn code is implemented much like urn, i.e., it deletes the context upon receiving zero-length buffer. Probably that works because (unlike urn) Squid performs asn requests on its behalf and does not send replies to client.

Reply:
> Are you implying that you saw zero length calls in your testing?

Sure, for otherwise, I would not notice that problem.

>  Did those calls have STORE_PENDING or STORE_OK entries?

IIRC they were not STORE_PENDING (but I need to re-test to confirm).
In that test, the zero-size message was the last one, so if the store_status were STORE_PENDING, my test would fail because it would not proceed further in urnHandleReply(). Probably it would even fall into an infinite recursion: urnHandleReply() schedules storeClientCopy(.., urnHandleReply,) and so on.

Reply:
I agree that you probably saw STORE_OK. Let's proceed with adding a Must() discussed in my previous comment then.